doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,294 @@
+Materialize world of determinism and consistency.
+
+The proposal here, then, is to bake the notion of volatility into the system. We
+start by marking volatile sources as such. Any view that depends on such a
+source becomes volatile. Any sink created from a volatile source or view becomes
+volatile. The optimizer will need to be careful not to treat different
+instantiations of a volatile source or view as logically equivalent. Exactly
+once sinks and other features that only work on nonvolatile sources can look
+for and reject volatile sources.
+
+### Design
+
+To start, I propose that we observe a simple rule:
+
+  * All PubNub, SSE, and Kinesis sources are marked as volatile.
+  * All other sources are not marked as volatile.
+
+In the future, I think we'll want to allow users to mark Kafka, S3, and file
+sources as volatile at their option, to capture when their *usage* of the source
+is volatile. See [Future work](#future-work) below.
+
+The only user-visible change at the moment will be an additional column in
+`SHOW SOURCES` that indicates whether a source is volatile
+
+```
+> SHOW SOURCES;
+ name        | type | volatile | materialized
+-------------+------+----------+-------------
+ pubnub_src  | user | true     | false
+ kafka_src   | user | false    | false
+```
+
+and similarly for `SHOW VIEWS`:
+
+```
+> SHOW VIEWS;
+ name         | type | volatile | materialized
+--------------+------+----------+-------------
+ pubnub_view  | user | true     | true
+ kafka_view   | user | false    | true
+```
+
+In the docs, we'll want to add a page about "Volatile sources" that explains
+what they are, why they are good for demos, and why they are bad for production
+applications that care about correctness. We can add a warning callout on
+PubNub, SSE, and Kinesis source pages that links to the volatile source page.
+
+The big "gotcha" with volatile sources is that different instantiations can have
+different data within the same `materialized` process. Consider a sequence like
+the following:
+
+```sql
+CREATE SOURCE market_orders_raw FROM PUBNUB SUBSCRIBE KEY 'sub-c-4377ab04-f100-11e3-bffd-02ee2ddab7fe' CHANNEL 'pubnub-market-orders';
+
+CREATE MATERIALIZED VIEW market_orders_1 AS
+  SELECT text::jsonb AS val FROM market_orders_raw;
+
+-- Imagine the user makes a cup of coffee now.
+
+CREATE MATERIALIZED VIEW market_orders_2 AS
+  SELECT text::jsonb AS val FROM market_orders_raw;
+```
+
+The data in `market_orders_1` and `market_orders_2` will not be equivalent,
+because they'll each have instantiated separate copies of the source at
+separate times. `market_orders_1` will contain some historical data that
+`market_orders_2` does not.
+
+The more I think about it, the more this seems like reasonable, even desirable
+behavior from a volatile source, provided we can educate users sufficiently.
+
+### Implications
+
+#### Exactly-once sinks
+
+Exactly-once sinks will need to error if they are created on a volatile source
```

Comment:
Some volatile sources will still have properties like "Materialize sees each message _at most once_", right? E.g., PubSub.

It might still be reasonable to allow creating sinks without a nonce from such a source -- although "exactly once" might be the wrong way to describe them.

This gives me an uneasy feeling that the one "volatile" bit isn't enough to describe the durability properties of sources with the richness that we need. What do you think?

Reply:
Yeah, I can imagine wanting a finer-grained set of bits in the future. I added a section on that. My thinking is that until you can write `VOLATILE` in `CREATE SOURCE`, we can screw around with our definition/understanding of volatility, and starting with a single bit seems like an easy way to get experience with this concept.